### json 불러와서 저장하기

In [1]:
import numpy as np
import pandas as pd
import json
import jpype
import glob
from random import shuffle
from bs4 import BeautifulSoup as bs

data_path = "data"

file_list = glob.glob("%s/*.json" % data_path)

json_train=[]

shuffle(file_list)
for json_file_name in file_list:
    print json_file_name
    json_file = json.loads(open(json_file_name).read())
    json_train += json_file["articles"]
    
print len(json_train)

data/3.json
data/11.json
data/16.json
data/6.json
data/12.json
data/9.json
data/1.json
data/14.json
data/0.json
data/10.json
data/4.json
data/13.json
data/2.json
data/7.json
data/5.json
data/15.json
data/8.json
data/17.json
172044


### json에서 author, author 형태소, forumid 추출

In [2]:
from konlpy.tag import Mecab

mecab = Mecab()

labeled_train = []

for cnt, article in enumerate(json_train):
    if cnt % 10000 == 0:
        print cnt

    labeled_train.append({
            "istroll": article["is_troll"],
            "author": article["author"],
            "author_pos": ["%s_%s" % (word, pos) for word, pos in mecab.pos(article["author"])],
            "forumid": article["forumid"],
        })

labeled_train = pd.DataFrame.from_dict(labeled_train)
labeled_train

0
10000
20000
30000
40000
50000
60000
70000
80000
90000
100000
110000
120000
130000
140000
150000
160000
170000


author  \
0                                   박멸   
1                                 rein   
2                                   촌년   
3                                 스캐치붘   
4                               미시간마치간   
5                                마인드보이   
6                                  유우나   
7                                  히트맨   
8                                 <동빠>   
9                               메디컬강냉이   
10                               명박x근혜   
11                               불량세스코   
12                                 지형태   
13                             홍어가풍년이노   
14                              메디컬강냉이   
15                              태생이전라도   
16                                카고15   
17                                flow   
18                             mariadb   
19                           국민에의한국민으로   
20                                바바토스   
21      닉네임글자수제한때문에끝까지못쓰는농민들은봐라닉이란건말이야   
22                               사랑의천사   
23                              자유인강산에   
24                               SATAN   
25                                  기멘   
26                                  세부   
27                               아라라기    
28                               Rocky   
29                              지금몇시야?   
...                                ...   
172014                     아이라인 그리는 판다   
172015                             박아희   
172016                  꾸꾸까까 ʕ·͡ᴥ·ʔ/ ♥   
172017                              정부   
172018                              정부   
172019                              정부   
172020                              정부   
172021                              정부   
172022                            酒袋飯囊   
172023                            미식헌터   
172024                        휴가중인 보도국   
172025                              정부   
172026                              정부   
172027                            뷰3유저   
172028                방해:AVI ʕ·͡ᴥ·ʔ/ ♥   
172029                             모히토   
172030                방해:AVI ʕ·͡ᴥ·ʔ/ ♥   
172031                 1붸뷉충아 아프지말고 죽어줘   
172032                             미모사   
172033                        pankkera   
172034                        휴가중인 보도국   
172035              99 bottles of beer   
172036                            doob   
172037                             양태호   
172038                            함초롬s   
172039                방해:AVI ʕ·͡ᴥ·ʔ/ ♥   
172040                            뷰3유저   
172041                            맛난호빵   
172042              ↑↓둘이커플, 중매하는 레이니가이   
172043                           Comea   

                                               author_pos forumid istroll  
0                                                [박멸_NNG]     잡담밭   False  
1                                               [rein_SL]     잡담밭   False  
2                                                [촌년_NNG]     잡담밭   False  
3                                        [스캐치_NNP, 붘_NNG]     잡담밭   False  
4                          [미_NNP, 시간_NNG, 마치_MAG, 간_NNG]     잡담밭   False  
5                                       [마인드_NNG, 보이_NNG]     잡담밭   False  
6                                               [유우나_NNP]     잡담밭   False  
7                                               [히트맨_NNP]   정치사회밭    True  
8                              [<_SY, 동_NNP, 빠_NNP, >_SY]     잡담밭   False  
9                                      [메디컬_NNG, 강냉이_NNG]     잡담밭   False  
10                           [명박_NNG, x_SL, 근_NNG, 혜_NNG]     잡담밭   False  
11                                      [불량_NNG, 세스코_NNP]     잡담밭   False  
12                                         [지_MM, 형태_NNG]     잡담밭   False  
13                   [홍어_NNG, 가_JKS, 풍년_NNG, 이_VCP, 노_EC]     잡담밭    True  
14                                     [메디컬_NNG, 강냉이_NNG]     잡담밭   False  
15                            [태생_NNG, 이전_NNG, 라도_VCP+EC]     잡담밭   False  
16                                        [카고_NNP, 15_SN]     유머밭    True  
17                                              [flow_SL]  

### forumid와 author을 label encoder로 숫자로 바꿔줌

In [3]:
from sklearn import preprocessing

le = preprocessing.LabelEncoder()

labeled_train["forumid"] = le.fit_transform(labeled_train["forumid"])
labeled_train["author"] = le.fit_transform(labeled_train["author"])

labeled_train

author                                         author_pos  forumid  \
0        13188                                           [박멸_NNG]       31   
1         3572                                          [rein_SL]       31   
2        23435                                           [촌년_NNG]       31   
3        16546                                   [스캐치_NNP, 붘_NNG]       31   
4        12677                     [미_NNP, 시간_NNG, 마치_MAG, 간_NNG]       31   
5        11624                                  [마인드_NNG, 보이_NNG]       31   
6        20033                                          [유우나_NNP]       31   
7        26838                                          [히트맨_NNP]       32   
8          489                         [<_SY, 동_NNP, 빠_NNP, >_SY]       31   
9        11997                                 [메디컬_NNG, 강냉이_NNG]       31   
10       12093                       [명박_NNG, x_SL, 근_NNG, 혜_NNG]       31   
11       14509                                  [불량_NNG, 세스코_NNP]       31   
12       22809                                     [지_MM, 형태_NNG]       31   
13       26480               [홍어_NNG, 가_JKS, 풍년_NNG, 이_VCP, 노_EC]       31   
14       11997                                 [메디컬_NNG, 강냉이_NNG]       31   
15       24604                        [태생_NNG, 이전_NNG, 라도_VCP+EC]       31   
16       23834                                    [카고_NNP, 15_SN]       27   
17        2727                                          [flow_SL]       31   
18        3234                                       [mariadb_SL]       31   
19        5902         [국민_NNG, 에_JKB, 의한_VV+ETM, 국민_NNG, 으로_JKB]       31   
20       13032                                   [바바_NNP, 토스_NNG]        4   
21        8944  [닉네임_NNG, 글자_NNG, 수제_NNG, 한_XSA+ETM, 때문_NNB, 에...       31   
22       15079                            [사랑_NNG, 의_JKG, 천사_NNG]       31   
23       21576                           [자유인_NNG, 강산_NNG, 에_JKB]       31   
24        1843                                         [SATAN_SL]       31   
25        6456                                       [기_VV, 멘_EC]       31   
26       15832                                           [세부_NNG]       31   
27       17254                                   [아라라_NNP, 기_ETN]       12   
28        1817                                         [Rocky_SL]       31   
29       22671              [지금_MAG, 몇_MM, 시_NNB, 야_VCP+EF, ?_SF]       31   
...        ...                                                ...      ...   
172014   17442                   [아이라인_NNG, 그리_VV, 는_ETM, 판다_NNG]       31   
172015   13230                                    [박_NNP, 아희_NNG]       34   
172016    7078  [꾸꾸_MAG, 까까_NNG, ʕ_SY, ·_SC, ͡ᴥ_SY, ·_SC, ʔ_SY...       14   
172017   22011                                           [정부_NNG]       12   
172018   22011                                           [정부_NNG]       31   
172019   22011                                           [정부_NNG]       31   
172020   22011                                           [정부_NNG]       31   
172021   22011                                           [정부_NNG]        4   
172022    4591                              [酒_XSN, 袋_SH, 飯囊_NNG]       31   
172023   12681                         [미식_NNG, 헌_XSV+ETM, 터_NNB]       31   
172024   26685                [휴가_NNG, 중_NNB, 인_VCP+ETM, 보도국_NNG]       31   
172025   22011                                           [정부_NNG]       31   
172026   22011                                           [정부_NNG]        5   
172027   14581                              [뷰_NNP, 3_SN, 유저_NNG]        1   
172028   13482  [방해_NNG, :_SC, AVI_SL, ʕ_SY, ·_SC, ͡ᴥ_SY, ·_SC...       31   
172029   12213                                     [모_IC, 히토_NNP]        2   
172030   13482  [방해_NNG, :_SC, AVI_SL, ʕ_SY, ·_SC, ͡ᴥ_SY, ·_SC...       31   
172031     209  [1_SN, 붸뷉충아_UNKNOWN, 아프_VA, 지_EC, 말_VX, 고_EC, ...       31   
172032   12648                                          [미모사_NNG]       32   
172033    3447                

###bag of words

In [4]:
from sklearn.feature_extraction.text import CountVectorizer

def make_bag_of_words(labeled_train, max_features, col_name):
    vectorizer = CountVectorizer(analyzer = "word", tokenizer = None, preprocessor = None,
                                stop_words = None, max_features=max_features)
    
    sentences = labeled_train[col_name].apply(lambda x:" ".join(x))
    train_data_features = vectorizer.fit_transform(sentences).toarray()

    col = ["bow_%s_%s" % (col_name, data) for data in vectorizer.get_feature_names()]
    df_bow = pd.DataFrame(train_data_features, columns = col, index=labeled_train.index)
    
    labeled_train = pd.concat([labeled_train, df_bow],axis=1)
    
    return labeled_train

##LDA

In [5]:
import gensim
from gensim import corpora, models
from gensim.corpora import TextCorpus, MmCorpus, Dictionary
from nltk.corpus import stopwords
from optparse import OptionParser

def make_lda(train, keep_n, num_topics, col_name):

    data = train[col_name]

    dictionary = corpora.Dictionary(data)
    dictionary.filter_extremes(keep_n=keep_n)

    corpus = [dictionary.doc2bow(text) for text in data]

    print("Make Lda..")

#    lda = gensim.models.ldamodel.LdaModel(corpus=corpus, id2word=dictionary, num_topics=num_topics, chunksize=1000, passes=1,distributed = True)
    lda = gensim.models.ldamodel.LdaModel(corpus=corpus, id2word=dictionary, num_topics=num_topics, chunksize=1000, passes=1)
    
    num = len(train)    
    df = []
    
    for i in range(0,num):
        if i % 10000 == 0:
            print(i)
        
        temp = [i[1] for i in lda.get_document_topics(corpus[i],minimum_probability=0)]
        #temp = max(lda.get_document_topics(corpus[i]),key=lambda item:item[1])[0]
        df.append(temp)
    
    col = ["lda_%s_%d" % (col_name, data) for data in range(0, num_topics)]
    df = pd.DataFrame(df, columns = col)
    df.index = train.index
    
    train = pd.concat([train, df], axis=1)
        
    return train

In [6]:
max_features = 1000
%time labeled_train = make_lda(labeled_train, 1000, 200, "author_pos")
#%time labeled_train = make_bag_of_words(labeled_train, 1000, "author_pos")

Make Lda..
0
10000
20000
30000
40000
50000
60000
70000
80000
90000
100000
110000
120000
130000
140000
150000
160000
170000
CPU times: user 2min 34s, sys: 2.21 s, total: 2min 36s
Wall time: 2min 37s


### predictor, model 생성

In [7]:
from sklearn.ensemble import RandomForestClassifier

label = 'istroll'
pre = labeled_train.columns.drop(['author_pos', 'author', label])
#pre = ['author','forumid']

model = RandomForestClassifier(n_estimators=10, n_jobs=6)

###cross validation

In [8]:
from sklearn import cross_validation

cv_value = 3

scores = cross_validation.cross_val_score(model, labeled_train[pre], labeled_train[label], cv=cv_value, scoring="roc_auc")
cv_result = scores.mean()

print(cv_result)

0.772313052394


## Result

 * author를 Label Encoder로 변환: 0.67331271058
 * author의 morphs를 Bag of Words(feature = 1,000 개)로 변환: 0.733333689652
 * author의 pos를 Bag of Words(feature = 1,000 개)로 변환: 0.769833271386
 * 데이터를 shuffle하고 pos를 Bag of Words(feautre = 1,000개)로 변환: 0.81172740086
 
 
 * author의 pos를 LDA로 변환 (num_topics = 200, keep_n = 1,000) : 0.748718319834
 * 데이터를 shuffle하고 author의 pos를 LDA로 변환 (num_topics = 200, keep_n = 1,000) : 0.772313052394

In [13]:
import pickle

pre = labeled_train.columns.drop(['author_pos', 'author'])

pickle.dump(labeled_train[pre], open("author, forum_%d.p" % max_features, "wb"), protocol = pickle.HIGHEST_PROTOCOL)

author                                         author_pos  forumid  \
0        15848                                   [세부_NNG, 이장_NNG]       16   
1         5653                                          [관리자_NNG]       31   
2         3752                                        [snoopy_SL]       31   
3         1258                                      [ISCARIOT_SL]       32   
4          136                                          [1234_SN]       23   
5        26299                                         [호놀룰루_NNP]       32   
6        18849                                    [영_NNG, 양죽_NNG]       31   
7        18849                                    [영_NNG, 양죽_NNG]       23   
8         2693                                      [fakebook_SL]       31   
9         6140                                           [그나_MAJ]       31   
10        7151                                     [뀨뀨꺄꺄_UNKNOWN]       32   
11        7151                                     [뀨뀨꺄꺄_UNKNOWN]        4   
12       11105                                    [로동_NNG, 자_XSN]       13   
13       17979              [야레_NNP, 바데_NNP, 키_NNG, 루_JKB, 코_NNG]       31   
14        4295                                [ㅁ_NNG, ㅇㅎ_UNKNOWN]        4   
15        2691                                [f_SL, 1_SN, ve_SL]       31   
16        2691                                [f_SL, 1_SN, ve_SL]       13   
17        2691                                [f_SL, 1_SN, ve_SL]       13   
18       15471                                           [새부_NNG]       31   
19        2691                                [f_SL, 1_SN, ve_SL]        4   
20        2691                                [f_SL, 1_SN, ve_SL]       31   
21       12761                      [민영_NNG, 민영_NNG, 하_XSV, 게_EC]       31   
22       21364                               [일_NR, 워_NNG, 충_NNG]       32   
23        2691                                [f_SL, 1_SN, ve_SL]        4   
24        7151                                     [뀨뀨꺄꺄_UNKNOWN]       12   
25       11226                                           [루시_NNP]       32   
26        7151                                     [뀨뀨꺄꺄_UNKNOWN]       32   
27       25915                                         [할렐루야_NNG]       32   
28       14295                                           [봉달_NNG]       31   
29       17541                                    [아카_NNP, 링_JKO]       12   
...        ...                                                ...      ...   
172014   22925                                   [질소_NNG, 까스_NNP]       31   
172015   10430                                   [두더지_NNG, 뿅_MAG]       26   
172016    1008                                     [EkardeNah_SL]       31   
172017   24288                                    [쿠데_NNG, 리_NNG]        5   
172018   12216                          [목동_NNP, 교주_NNP, 박병호_NNP]       26   
172019   12254                             [못난이_NNG, 3_SN, 째_XSN]       17   
172020     663                                           [Bee_SL]       31   
172021   15076  [사랑_NNG, 은_JX, 옆_NNG, 에서_JKB, 함께_MAG, 하_VV, 는_...       32   
172022   16593                            [스티브_NNP, 짭새_NNG, ™_SY]       31   
172023   10045                                     [덧_XPN, 글_NNG]       31   
172024   16593                            [스티브_NNP, 짭새_NNG, ™_SY]       31   
172025   20523    [이_MM, 사이트_NNG, 마음_NNG, 에_JKB, 든_VV+ETM, 닭_NNG]        7   
172026   16593                            [스티브_NNP, 짭새_NNG, ™_SY]       31   
172027   26811               [히_VV, 드_EC, 랍_NNG, 더_MAG, 볼_VV+ETM]       26   
172028   26811               [히_VV, 드_EC, 랍_NNG, 더_MAG, 볼_VV+ETM]       26   
172029   18868                          [영_NNG, 할_XSV+ETM, 트_NNG]       32   
172030   17496                                 [아일랜드_NNP, 남자_NNG]       31   
172031   22675  [지금_NNG, 부터_JX, 렙_NNG, 을_JKO, 한다_VV+EC, 농_NNG,...       27   
172032   16593                            [스티브_NNP, 짭새_NNG, ™_SY]       31   
172033    3346                